Realm/Realm/Realm.cs

```diff
@@ -1357,6 +1344,54 @@ internal void ExecuteOutsideTransaction(Action action)
 
         #endregion Transactions
 
+        internal class SessionProvider
+        {
+            private WeakReference<Session>? _weakSessionRef;
+            private Session? _strongSessionRef;
+
+            public Session GetSession(SharedRealmHandle handle)
+            {
+                if(_strongSessionRef?.IsClosed == false)
+                {
+                    return _strongSessionRef;
+                }
+                else if (_weakSessionRef?.TryGetTarget(out var targetSession) == true && !targetSession.IsClosed)
+                {
+                    return targetSession;
+                }
+
+                var session = new Session(handle.GetSession());
+                session.Subscribed += Session_Subscribed;
+                session.Unsubscribed += Session_Unsubscribed;
+
+                if (_weakSessionRef is null)
+                {
+                    _weakSessionRef = new WeakReference<Session>(session);
+                }
+                else
+                {
+                    _weakSessionRef.SetTarget(session);
+                }
+
+                return session;
+            }
+
+            private void Session_Subscribed(object? sender, EventArgs e)
+            {
+                if (sender is not Session session)
+                {
+                    return;
+                }
+
+                _strongSessionRef = session;
+            }
+
+            private void Session_Unsubscribed(object? sender, EventArgs e)
+            {
+                _strongSessionRef = null;
+            }
```

Comment:
Rather than use events, we can pass the provider to the session constructor and have the session call methods on the provider. I know it's roughly the same thing, but I feel it's more straightforward.

Reply:
The constructor is used also here : https://github.com/realm/realm-dotnet/blob/68febca6905095940b22bcee1d7f58e57aa2910b/Realm/Realm/Handles/SessionHandle.cs#L293

So I think it would be awkward to pass a nullable provider. 
Even not considering that, I am not totally convinced about passing the provider as an argument to the constructor, because it means that now `Session` would have a dependency on `SessionProvider`, that is not necessary.

Reply:
Fair enough. I do feel the session already has a dependency on the `SessionProvider`, but we're trying to be roundabout about it. These events don't make a ton of sense by themselves and we're only raising them because we know there's going to be some very special class that cares about whether the session has subscribers or not. Since we're modifying the Session API to cater to the new provider class we're introducing and we don't expect these events to ever be used by anything else, I'd say we have a pretty strong coupling between those two types. That being said, it's up to you to decide on the design - I'd be fine with either approach.

Reply:
I see your point. I changed to pass a delegate to the constructor, it feels like a middle-ground, so we don't need to make the methods on the provider public.